---
title: "RCPA: a R package for consensus pathway analysis and visualization"
author:
output:
  html_document:
    df_print: paged
    toc: true

editor_options:
  chunk_output_type: inline
csl: ieee.csl
bibliography: ref.bib
vignette: |
  %\VignetteIndexEntry{RCPA} 
  %\usepackage[UTF-8]{inputenc}
  %\VignetteEngine{knitr::rmarkdown} 
---
# Introduction
Consensus pathway analysis is a powerful bioinformatics method used to integrate diverse types of biological data and identify biological pathways that are significantly altered in a specific experimental condition. This approach has been successfully applied to a wide range of research areas, including cancer, neuroscience, and immunology. We present RPCA, a R's package for performing consensus pathway analysis using publicly available tools and resources. Here, we present a standalone R's package for consensus pathway analysis. We show the key steps involved in the analysis: (i) data acquisition from Gene Expression Omnibus, (ii) differential expression analysis using limma, edgeR, and DESeq2, (iii) genesets acquisition from KEGG and GO databases, (iv) pathway enrichment analysis using ORA, FGSEA, and GSA methods, (v) meta-analysis to produce consensus scoring using Fisher, Stouffer, addCLT, GeoMean, and minP, and (vi) and genes/pathways visualization. We also provide guidelines for selecting appropriate data sources and statistical thresholds to ensure robust and reproducible results. 

# Data acquisition
RCPA provides one-stop solution for downloading and processing high-throughput gene expression data from Gene Expression Omnibus (GEO) using the function *downloadGEO*. This function is specifically designed to support both Affymetrix and Agilent protocols, allowing you to seamlessly access and manipulate datasets from these popular microarray technologies.
The simple interface allow easy download, preprocess, and analyze data from GEO, regardless of user experience level. 

Users need to provide four parameters into *downloadGEO* function to download the data. First, we need an accession number which is a unique identifier assigned to a dataset deposited in the GEO database. It consists of a combination of letters and numbers that uniquely identify a specific dataset. The accession number is typically preceded by the letters ``GSE'' and can be found in the title of the dataset and on the data summary page in the GEO database. We will use dataset with accession number *GSE48350* [@berchtold2008gene] for demonstration. Second, users need to provide platform keyword which refers to a specific microarray platform, or chip, that is used for high-throughput gene expression profiling (e.g. hgu133a.db. Third, the *protocol* can be selected between *affymetrix* and *agilent*. Finally, a specific path should be set to *destDir*. The full script download the data can be presented as:

```{r, eval=FALSE}
data <- downloadGEO(GEOID = "GSE33970", platform = "GPL570", protocol = "affymetrix", destDir = "./data/")
```


RCPA package uses widely-adopted *SummarizedExperiment* object as its standard output format. By utilizing this robust and flexible data structure, the RCPA package ensures the workflow to integrate smoothly with various upstream and downstream tools available in many public repositories such as CRAN and Bioconductor. The *SummarizedExperiment* object is structured in a way that captures the various dimensions and metadata associated with the data, making it an ideal choice for storing and manipulating complex datasets.

The structure of a *SummarizedExperiment* object consists of three primary components: (i) *Assays* store the actual experiment data, such as gene expression values or read counts. *Assays* are organized as matrices or similar data structures, with rows representing features (e.g., genes or transcripts) and columns representing samples or observations. In cases where multiple data types or normalization methods are used, the *SummarizedExperiment* object can accommodate multiple assay matrices, (ii) *rowData* contains feature-level metadata, such as gene annotations, transcript information, or genomic coordinates and. *rowData* is typically stored as a DataFrame, with each row corresponding to a feature in the assay(s) and each column representing a specific metadata attribute and (iii) *colData* captures sample-level metadata, such as clinical information, experimental conditions, or batch information. Similar to *rowData*, *colData* is usually stored as a DataFrame, with each row corresponding to a sample in the assay(s) and each column representing a specific metadata attribute.

These three components are interconnected through the SummarizedExperiment object, ensuring that the assay data, rowData, and colData remain consistent and synchronized. By providing a unified data structure that can accommodate diverse biological experiments, the SummarizedExperiment object simplifies data management and promotes reproducible research practices in the Bioconductor ecosystem.

# Differential expression analysis
Pathway analysis aims to identify the biological pathways or functional categories that are significantly enriched with genes that are associated with a particular phenotype or condition. In other words, pathway analysis seeks to understand the molecular mechanisms and biological processes that are involved in a particular phenotype or condition, such as a disease, by identifying the genes that are involved and their interactions within pathways. By analyzing the gene list within the context of known biological pathways or functional categories, pathway analysis can identify the pathways that are over-represented in the gene list, providing insight into the underlying biological processes and mechanisms that contribute to the phenotype or condition.

There are several ways to obtain a gene list for pathway analysis, depending on the research question, the type of data being analyzed, and the available resources. The common approach is differential gene expression analysis (DE) which involves comparing gene expression levels between two or more conditions or groups, such as treatment versus control or disease versus healthy. The genes that are significantly differentially expressed are then used for pathway analysis. Gene set analysis uses pre-defined sets of genes, such as gene ontologies or gene pathways, for pathway analysis. In this approach, the analysis is focused on the enrichment of genes within a pre-defined gene set rather than identifying specific differentially expressed genes. Another approach focuses on literature-based gene selection in which genes that have been reported in the literature to be associated with a particular biological process or disease. These genes can then be used for pathway analysis. Genomic annotation analysis uses genomic annotation data, such as chromatin state or DNA methylation, to identify genes that are epigenetically regulated and may be involved in a particular biological process or disease. The last approach uses network-based analysis that aims at constructing a gene co-expression network or protein-protein interaction network and selecting genes that are highly connected or central to the network.

RCPA package has a dedicated function *DE-analysis* to obtain gene list from omics data without knowing prior knowledge using DE approach. *DE-analysis* provides a wrapper function to use common DE methods including *limma* [@ritchie2015limma], *edgeR* [@robinson2010edger], and *DESeq2* [@love2014moderated]. In order to perform DE analysis, *DE-analysis* requires *summarizedExperiment* object, DE method selection (limma, DESeq2, edgeR), design matrix, contrast matrix and annotation database.

### Data preparation
A simple way to prepare a simulated data can be shown as follows:

```{r, eval=FALSE}
library(hgu133plus2.db)
library(AnnotationDbi)
library(SummarizedExperiment)
library(limma)
# generate a random gene expression matrix
set.seed(123)
exprs <- round(matrix(2^abs(rnorm(1000, sd = 4)), nrow = 100, ncol = 10))
rownames(exprs) <- sample(keys(hgu133plus2.db, keytype = "PROBEID"), nrow(exprs), replace = FALSE)
colnames(exprs) <- paste0("sample", 1:10)

controlSamples <- paste0("sample", 1:5)
conditionSamples <- paste0("sample", 6:10)

colData <- data.frame(
    row.names = colnames(exprs),
    group = factor(c(rep("control", length(controlSamples)), rep("condition", length(conditionSamples)))),
    pair = factor(c(seq_along(controlSamples), seq_along(conditionSamples)))
)

summarizedExperiment <- SummarizedExperiment(
    assays = list(counts = exprs),
    colData = colData
)
```
### Design matrix
A design matrix is a mathematical representation of the experimental design that describes the relationships between the samples and the variables or factors of interest. The design matrix is used in linear modeling approaches, such as the ones implemented in the R packages *limma*, *edgeR* and *DESeq2*, to identify differentially expressed genes between different groups or conditions. In a design matrix, each row represents an individual sample (e.g., RNA-seq or microarray data from a specific biological replicate), and each column represents a variable or factor in the experiment, such as different treatments, time points, or tissue types. The matrix entries are usually binary (0 or 1), indicating whether a specific factor is present (1) or absent (0) in the corresponding sample. A simple design matrix can be constructed as follows:

```{r, eval=FALSE}
# control vs condition
design <- model.matrix(~0 + group, data = colData)
```

The design matrix can be shown as follows:
```
         groupcondition groupcontrol
sample1               0            1
sample2               0            1
sample3               0            1
sample4               0            1
sample5               0            1
sample6               1            0
sample7               1            0
sample8               1            0
sample9               1            0
sample10              1            0
attr(,"assign")
[1] 1 1
attr(,"contrasts")
attr(,"contrasts")$group
[1] "contr.treatment"
```
### Contrast matrix
A contrast matrix is typically a matrix with one or more rows and columns equal to the number of variables or factors in the design matrix. Each row of the contrast matrix represents a specific comparison of interest (e.g., treatment vs. control, time point A vs. time point B, etc.). The entries in the matrix indicate the weights applied to each factor or coefficient when making the comparison. These weights are usually 1, -1, or 0, with 1 representing the factor of interest, -1 representing the reference or baseline condition, and 0 indicating that a factor is not involved in the comparison. The contrast matrix is used to extract the relevant information from the fitted linear model by multiplying it with the estimated coefficients. This results in a set of estimated differences in gene expression between the specified conditions, which can then be tested for statistical significance to identify differentially expressed genes. We can create a contrast matrix using *makeContrasts* function available in *limma* package. 

```{r, eval=FALSE}
contrast <- makeContrasts("groupcondition-groupcontrol", levels = design)
```
The contrast matrix can be shown as follows:

```
                Contrasts
Levels           groupcondition-groupcontrol
  groupcondition                           1
  groupcontrol                            -1
```

### Gene annotation

If the users do not want perform genes ID mapping for the assay, the *annotation* parameter can be set to *NULL*. Otherwise, a biological annotation databases should be provide. Annotation databases contain information about genes, transcripts, proteins, or other genomic features, such as gene symbols, descriptions, chromosomal locations, or functional annotations (e.g., Gene Ontology terms). The AnnotationDbi package is particularly useful in the context of high-throughput genomic data analysis, such as microarray or RNA-seq experiments, as it allows researchers to easily map and retrieve annotation information for a list of genes or other genomic features. A sample annotation data frame can be shown as follows:

```
            FROM        TO
1      1007_s_at       780
2        1053_at      5982
3         117_at      3310
4         121_at      7849
5      1255_g_at      2978
6        1294_at      7318
7        1316_at      7067
8        1320_at     11099
9      1405_i_at      6352
10       1431_at      1571
11       1438_at      2049
12       1487_at      2101
13     1494_f_at      1548
14  1552256_a_at       949
15  1552257_a_at     23170
16    1552261_at     10406
17    1552263_at      5594
18  1552264_a_at      5594
19    1552266_at    203102
20    1552269_at    128153
```
A sample script to perform DE analysis without genes ID mapping in RCPA package can be seen as follows:

```{r, eval=FALSE}
DERes <- runDEAnalysis(summarizedExperiment, method = "DESeq2", design, contrast, annotation = NULL)
```

The result of DE analysis is stored in the *SummarizedExperiment* object. To access the DE analysis result, we can call *rowData* function. DE result is a data frame in which rows are genes and columns are probe IDs, gene IDs, and statistical result. The full data frame can be seen as:
```
               PROBEID          ID     p.value statistic     logFC dispersion       pFDR
           <character> <character>   <numeric> <numeric> <numeric>  <numeric>  <numeric>
9356       210343_s_at        9356 4.81682e-05  -4.06434  -5.76835    1.41926 0.00390163
1833         206439_at        1833 2.46935e-04  -3.66542  -3.88897    1.06099 0.01000085
4801         244704_at        4801 9.21429e-04   3.31348   4.03540    1.21787 0.01967807
54434     1554274_a_at       54434 9.71756e-04  -3.29858  -4.75052    1.44017 0.01967807
81552      208091_s_at       81552 1.39365e-03  -3.19596  -4.12387    1.29034 0.02105996
```
# Genesets and pathways
Pathways and gene sets are essential concepts in systems biology, as they provide a framework for understanding the organization and interactions of these molecular components within the cell. By analyzing pathways and gene sets, researchers can gain valuable insights into the biological processes and molecular mechanisms underlying various diseases, conditions, and cellular responses. Gene sets are collections of genes that share a common biological function, cellular component, or molecular process. These sets can be derived from various sources, including curated pathway databases, gene ontology annotations, or experimental data. By grouping genes based on their functional roles, gene sets provide a simplified and structured framework for interpreting large-scale gene expression data. 

Gene Ontology (GO) [@ashburner2000gene] and the Kyoto Encyclopedia of Genes and Genomes (KEGG) [@kanehisa2000kegg] pathway databases are two essential resources that provide structured frameworks for annotating and analyzing genes, proteins, and other molecular components. By offering standardized vocabularies and curated information on molecular functions, cellular components, biological processes, and pathways, these databases enable researchers to systematically explore the relationships between genes and their functional roles in various biological systems. RCPA package provides a function *getGeneSets* to obtain genesets from two databases

## Obtaining genesets from KEGG

To obtain genesets from KEGG, user need to provide the keyword for organism (e.g "hsa" for human). The full supporting list is avaialble at https://www.genome.jp/kegg/catalog/org_list.html. The script to download genelist from KEGG is as follows:

```{r, eval=FALSE}
gensets <- getGeneSets("KEGG", "hsa", minSize = 10, maxSize = 1000)
```

The output is a list that contais database name (KEGG), list of geneset for each pathway, and a vector contain description of each pathway. A typical pathway and gene list can be seen as follows:

```
$`path:hsa05204`
 [1] "10"        "100861540" "10720"     "10941"     "119391"    "1543"      "1544"      "1545"      "1548"      "1549"     
[11] "1551"      "1553"      "1557"      "1558"      "1559"      "1562"      "1571"      "1576"      "1577"      "1646"     
[21] "2052"      "221357"    "27306"     "2938"      "2939"      "2940"      "2941"      "2944"      "2946"      "2947"     
[31] "2948"      "2949"      "2950"      "2952"      "2953"      "3290"      "373156"    "374875"    "4257"      "4258"     
[41] "4259"      "445329"    "54490"     "54575"     "54576"     "54577"     "54578"     "54579"     "54600"     "54657"    
[51] "54658"     "54659"     "5743"      "574537"    "64816"     "653689"    "6799"      "6817"      "6818"      "6822"     
[61] "7363"      "7364"      "7365"      "7366"      "7367"      "79799"     "873"       "9"         "9446"     

$`path:hsa05205`
  [1] "10000"  "1026"   "10451"  "10818"  "10855"  "11211"  "117581" "1277"   "1278"   "1432"   "1499"   "1514"   "1634"  
 [14] "1655"   "1839"   "1956"   "1975"   "2002"   "2017"   "2064"   "2065"   "2066"   "207"    "208"    "2099"   "2247"  
 [27] "2260"   "22800"  "22808"  "2316"   "2317"   "2318"   "2335"   "23365"  "2475"   "2535"   "2549"   "2719"   "27250" 
 [40] "2817"   "286"    "287"    "288"    "2885"   "29102"  "3059"   "3082"   "3091"   "3236"   "3265"   "3316"   "3339"  
 [53] "3479"   "3480"   "3481"   "3549"   "355"    "356"    "3593"   "3673"   "3678"   "3685"   "3688"   "369"    "3690"  
 [66] "3693"   "3708"   "3709"   "3710"   "3791"   "3845"   "387"    "388112" "4060"   "406902" "406903" "406991" "4087"  
 [79] "4193"   "4233"   "4313"   "4318"   "4478"   "4609"   "4659"   "4660"   "4893"   "5058"   "51196"  "51384"  "5170"  
 [92] "5290"   "5291"   "5293"   "5295"   "5296"   "5328"   "5329"   "5335"   "5336"   "54361"  "54776"  "5499"   "5500"  
[105] "5501"   "5566"   "5567"   "5568"   "5578"   "5579"   "5582"   "5594"   "5595"   "5600"   "5603"   "5604"   "5605"  
[118] "5727"   "5747"   "5777"   "5781"   "5829"   "5879"   "5894"   "595"    "5962"   "60"     "60495"  "6093"   "6194"  
[131] "6198"   "6199"   "6237"   "6300"   "6382"   "6383"   "6385"   "6469"   "6548"   "6608"   "6654"   "6655"   "6714"  
[144] "673"    "6774"   "7023"   "7040"   "7042"   "7057"   "7074"   "7078"   "7097"   "7099"   "71"     "7124"   "7157"  
[157] "7291"   "7409"   "7410"   "7422"   "7430"   "7448"   "7471"   "7472"   "7473"   "7474"   "7475"   "7476"   "7477"  
[170] "7478"   "7479"   "7480"   "7481"   "7482"   "7483"   "7484"   "7855"   "7976"   "79923"  "80326"  "81029"  "815"   
[183] "816"    "817"    "818"    "8321"   "8322"   "8323"   "8324"   "8325"   "8326"   "836"    "84309"  "8503"   "857"   
[196] "858"    "859"    "867"    "8826"   "89780"  "9138"   "9475"   "960"    "967"    "998"   
```

Example of pathway description is shown as follows:
```
                                                            path:hsa00010 
                                           "Glycolysis / Gluconeogenesis" 
                                                            path:hsa00020 
                                              "Citrate cycle (TCA cycle)" 
                                                            path:hsa00030 
                                              "Pentose phosphate pathway" 
                                                            path:hsa00040 
                               "Pentose and glucuronate interconversions" 
                                                            path:hsa00051 
                                        "Fructose and mannose metabolism" 
                                                            path:hsa00052 
                                                   "Galactose metabolism" 
                                                            path:hsa00053 
                                      "Ascorbate and aldarate metabolism" 
                                                            path:hsa00061 
                                                "Fatty acid biosynthesis" 
                                                            path:hsa00062 
                                                  "Fatty acid elongation" 
                                                            path:hsa00071 
                                                 "Fatty acid degradation" 
                                                            path:hsa00100 
                                                   "Steroid biosynthesis" 
                                                            path:hsa00120 
                                         "Primary bile acid biosynthesis" 
                                                            path:hsa00130 
```

## Obtaining genesets from GO

To obtain genesets from GO, user need to provide taxonomy IDs and namespace. Taxonomy IDs are used to classify and organize species within a hierarchical system that reflects their evolutionary relationships. In GO, taxonomy IDs are used to associate specific GO terms with the organisms to which they apply. This is important because certain biological processes, molecular functions, or cellular components may be unique to, or have a different context in, specific organisms or groups of organisms.Namespace refers to the three distinct categories or domains under which GO terms are organized. Each GO term belongs to one of the following three namespaces:

* Biological Process (BP): This namespace encompasses terms that describe a series of molecular events or a pathway, with a defined beginning and end, that contribute to a specific biological outcome or function. Examples include cell cycle, DNA repair, or apoptosis.

* Molecular Function (MF): This namespace includes terms that describe the biochemical activities or actions of gene products at the molecular level, such as catalytic or binding activities. Examples include protein kinase activity, DNA binding, or ion channel activity.

* Cellular Component (CC): This namespace consists of terms that describe the locations within a cell or the extracellular environment where gene products are active or found. These locations can be cellular structures, such as organelles or protein complexes, or broader regions like the cell membrane. Examples include nucleus, mitochondrion, or the proteasome complex.

The example script to get genesets from GO database is as follows:

```{r, eval=FALSE}
gensets <- getGeneSets("GO", taxid = 9606, namespace = "biological_process", minSize = 10, maxSize = 1000)
```

The output is a list that contais database name (GO), list of geneset for each term, and a named vector contain description of each term A typical term and gene list can be seen as follows:

```
$`GO:0048511`
 [1] "983"    "1020"   "1387"   "1390"   "1628"   "1642"   "1815"   "2146"   "6096"   "6478"   "6667"   "7153"   "7874"  
[14] "8242"   "8451"   "9475"   "10533"  "11335"  "22871"  "23291"  "26223"  "26224"  "54551"  "55269"  "55294"  "57805" 
[27] "84667"  "85457"  "92369"  "94233"  "150094"

$`GO:0048514`
 [1] "1000" "1282" "1545" "1948" "2116" "2321" "2324" "3397" "5921" "6654" "8456"

$`GO:0048538`
 [1] "673"   "1399"  "2255"  "2304"  "2625"  "3200"  "3280"  "6495"  "6647"  "6722"  "8379"  "8543"  "9133"  "9464"  "10269"
[16] "57864"

$`GO:0048565`
 [1] "887"   "1281"  "1543"  "2263"  "5925"  "25836" "55366" "57680" "57680" "93649"
```

Example of term description is shown as follows:
```
                                                    GO:0045739 
                                                                "positive regulation of DNA repair" 
                                                                                         GO:0045740 
                                                           "positive regulation of DNA replication" 
                                                                                         GO:0045741 
                       "positive regulation of epidermal growth factor-activated receptor activity" 
                                                                                         GO:0045742 
                        "positive regulation of epidermal growth factor receptor signaling pathway" 
                                                                                         GO:0045746 
                                                   "negative regulation of Notch signaling pathway" 
                                                                                         GO:0045747 
                                                   "positive regulation of Notch signaling pathway" 
                                                                                         GO:0045765 
                                                                       "regulation of angiogenesis" 
                                                                                         GO:0045766 
                                                              "positive regulation of angiogenesis" 
                                                                                         GO:0045773 
                                                            "positive regulation of axon extension" 
                                                                                         GO:0045785 
                                                             "positive regulation of cell adhesion" 
                                                                                         GO:0045786 
                                                                "negative regulation of cell cycle" 
```


# Pathway enrichment analysis

Enrichment analysis (EA) is a technique used to derive biological insight from lists of significantly altered genes. The list of genes can be obtained from Differential Expression (DE) analysis or users’ interest. The EA methods rely on the knowledge databases (e.g. KEGG and GO) to identify biological pathways or terms that are enriched in a gene list more than would be expected by chance. The outcome of the EA would be the in-depth and contextualized findings to help understand the mechanisms of disease, genes and proteins associated with the etiology of a specific disease or drug target.

Over more than a decade, there are many methods have been developed for EA. RCPA package offers a wrapper function *runPathwayAnalysis* to perform enrichment analysis using three popular methods including Over Representation Analysis (ORA) [@khatri2002profiling], Fast Gene Set Enrichment Analysis (FGSEA) [@korotkevich2016fast], and Gene Set Analysis (GSA) [@efron2007testing].

# Meta-analysis

# Enriched pathway visualization
## Plot heatmap
## Ploy volcano
## Plot pathway bar charts

# Reference

